sky/provision/azure/instance.py

```diff
@@ -45,64 +141,412 @@ def get_azure_sdk_function(client: Any, function_name: str) -> Callable:
+        raise RuntimeError(
+            'The number of pending/running/stopped/stopping '
+            f'instances combined ({len(existing_instances)}) in '
+            f'cluster "{cluster_name_on_cloud}" is greater than the '
+            f'number requested by the user ({config.count}). '
+            'This is likely a resource leak. '
+            'Use "sky down" to terminate the cluster.')
+
+    to_start_count = config.count - len(pending_instances) - len(
+        running_instances)
+
+    if to_start_count < 0:
+        raise RuntimeError(
+            'The number of running+pending instances '
+            f'({config.count - to_start_count}) in cluster '
+            f'"{cluster_name_on_cloud}" is greater than the number '
+            f'requested by the user ({config.count}). '
+            'This is likely a resource leak. '
+            'Use "sky down" to terminate the cluster.')
+
+    if config.resume_stopped_nodes and to_start_count > 0 and (
+            stopping_instances or stopped_instances):
+        time_start = time.time()
+        if stopping_instances:
+            plural = 's' if len(stopping_instances) > 1 else ''
+            verb = 'are' if len(stopping_instances) > 1 else 'is'
+            # TODO(zhwu): double check the correctness of the following on Azure
+            logger.warning(
+                f'Instance{plural} {[inst.name for inst in stopping_instances]}'
+                f' {verb} still in STOPPING state on Azure. It can only be '
+                'resumed after it is fully STOPPED. Waiting ...')
+        while (stopping_instances and
+               to_start_count > len(stopped_instances) and
+               time.time() - time_start < _RESUME_INSTANCE_TIMEOUT):
+            inst = stopping_instances.pop(0)
+            per_instance_time_start = time.time()
+            while (time.time() - per_instance_time_start <
+                   _RESUME_PER_INSTANCE_TIMEOUT):
+                status = _get_instance_status(compute_client, inst,
+                                              resource_group)
+                if status == AzureInstanceStatus.STOPPED:
+                    break
+                time.sleep(1)
+            else:
+                logger.warning(
+                    f'Instance {inst.name} is still in stopping state '
+                    f'(Timeout: {_RESUME_PER_INSTANCE_TIMEOUT}). '
+                    'Retrying ...')
+                stopping_instances.append(inst)
+                time.sleep(5)
+                continue
+            stopped_instances.append(inst)
+        if stopping_instances and to_start_count > len(stopped_instances):
+            msg = ('Timeout for waiting for existing instances '
+                   f'{stopping_instances} in STOPPING state to '
+                   'be STOPPED before restarting them. Please try again later.')
+            logger.error(msg)
+            raise RuntimeError(msg)
+
+        resumed_instances = stopped_instances[:to_start_count]
```

Comment:
```suggestion
        instances_to_resume = stopped_instances[:to_start_count]
```
those are not resumed yet?

Reply:
Looks great!